### PR TITLE
initial invite api

### LINKED ISSUE #198: Create class that implements `invite` namespace API
## Description

We need to implement the client-level API specified in https://github.com/digidem/mapeo-core-docs/blob/main/client-docs/README.md#clientinvite. We currently have a `MapeoRPC` class that manages _sent_ invites but we need something that will handle _incoming_ invites and account for multiproject (note, not necessarily multi-peer). Instead of updating that implementation, we should implement another class that uses an rpc instance and is responsible for handling incoming invites and responding to them based on project. 

Some pseudocode for an imagined implementation:

```ts
// Create the class that manages these invites
class InviteApi {
  // Accepts a MapeoRPC instance
  constructor({ rpc })
    this.rpc = rpc

    // attach listener for 'invite' event, re-emit as 'invite-received'
    // Track project key, and encryption keys for invite.
    // dedupe - if we receive two invites for same project, 
    // *do* track the clientId that need to respond to, but *don't* emit twice
  }

  accept(projectId) {
    const clientIds = this.#getClientsThatSentInvite(projectId)
    for (const clientId of clientIds) {
        this.rpc.inviteResponse(clientId, { projectId, decision: 'ACCEPT' })
      }
    }

  reject(projectId) {
    // similar to above
  }
}
```

Based on initial discussion, minimal changes are needed in the `MapeoRPC` class implementation, which are being tracked in https://github.com/digidem/mapeo-core-next/issues/197 and #196 

## Tasks

- [ ] implement `InviteApi` class

### PR DESCRIPTION
**Pull request by sethvincent**
closes #198

### CODE REVIEW

File: src/invite-api.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+   * @param {string} options.projectId
+   * @param {InviteResponse_Decision} options.decision
+   */
+  #respond({ projectId, decision }) {
+    const peerIds = this.#getPeerIds(projectId)
+    const projectKey = idToKey(projectId)
+
+    // TODO: should this reply to  one peer with `ACCEPT` and the rest with `ALREADY`?
+    // How is the `ALREADY` decision determined? It looks like that isn't used yet.
+    // Does anything bad happen if we respond to multiple peers with `ACCEPT`?
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/invite-api.js:61]: Ah good question. The impact would be that each of the invitors receiving `ACCEPT` would write a role record for the invitee. This would be confusing to deal with - we would have duplicate records of a device being added to the project. We should handle this in case it does happen (I've added this to #189).

On reflection I think what would make sense would be to respond `ACCEPT` to one peer (the first that we received an invite from?) and send `ALREADY` to the others.

[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/invite-api.js:61]: In the case of peers disconnecting, we should send `ACCEPT` to the first still-connected peer that we find.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Looks good! I've added some responses to your questions. You'll need to coordinate with @achou11 for the API for looking up existing projects and joining a new one.

One issue to handle is disconnecting peers. We should emit an event for this, and return an error from `accept` or `reject` if trying to respond to a disconnected peer.

I think we should try not to add a project if we are unable to send the `ACCEPT` message, since we will not know if the invitor has actually written us to the auth core to actually access the project.

File: src/invite-api.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import { TypedEmitter } from 'tiny-typed-emitter'
+import { InviteResponse_Decision } from './generated/rpc.js'
+import { idToKey, keyToId } from './utils.js'
+
+/** @typedef {import('./rpc/index.js').MapeoRPC} MapeoRPC */
+
+export class InviteApi extends TypedEmitter {
+  // TODO: are invites persisted beyond this api?
```
[the remark this answers is not in the document]
Comment (achou11) [reply src/invite-api.js:8]: > This means that the InviteApi needs to look up which projects the device is already a member of - @achou11 can you suggest the best API for this? We'll need to pass something down to the InviteApi constructor.

Initial thought is to provide a callback in the constructor to determine membership for a project id e.g.

```ts
new InviteApi({
  // can update option name and return a string if there are more complex membership states to represent
  isMember: (projectId: string) => Promise<boolean> 
})
```

alternatively can pass a `sharedDb` option like we do for `MapeoProject` and let the InviteApi handle making the query itself. this approach may be useful if there are other project-related queries that may be needed

[the remark this answers is not in the document]
Comment (achou11) [reply src/invite-api.js:8]: with some additional pondering, thinking the sharedDb option is preferred for me, as it allows code specifically needed for the InviteApi to live with the implementation, as opposed to it leaking to the creator of the instance.

I'm sure my thoughts will bounce back and forth before I end up in this position again 😄

[the remark this answers is not in the document]
Comment (achou11) [reply src/invite-api.js:8]: just as I anticipated, seeing https://github.com/digidem/mapeo-core-next/pull/217#discussion_r1303951535 has me thinking about this again. 

So to sum it up, the InviteApi needs a way to do these things:

1. Check project membership
2. Add the project when accepting a received invite

Passing the `sharedDb` option would solve 1 well, but wouldn't solve 2. 

If we wanted to solve both, passing specific constructor callback options would be the simplest I guess i.e.

```ts
new InviteApi({
  isMember: ...,
  addProject: ...,
})
```

may be that's just the way to go, before I start thinking of some (probably) convoluted approach using events  😄

[the remark this answers is not in the document]
Comment (achou11) [reply src/invite-api.js:8]: currently using the callback-based approach in https://github.com/digidem/mapeo-core-next/pull/232/files#diff-a2f290f8510d711051a63f76d7d7b06721c1dd5b6ac7d86e8e3024795c2d829dR12, which I think i like right now.

Feels sensible that these API classes don't need to dive into the actual db and do their own queries

[the remark this answers is not in the document]
Comment (achou11) [reply src/invite-api.js:8]: leaning into the callback-based approach, maybe all of these API classes can have a singular constructor opt called `queries`, which is responsible for being the interface between the class and dbs created outside the class. 

For example, the InviteApi would have something like:

```ts
type InviteApiQueries = {
  isMember: ...
  addProject: ...
}
```

and instantiation in the `MapeoProject` will look like:

```ts
new InviteApi({
  queries: {
    isMember: async () => ...,
    addProject: async () => ...,
  },
  ...
})
```

honestly this is mostly aesthetic as it's not functionally different than specifying an opt for each callback. gut feeling is that it'd be a little cleaner to work with this, especially if an API class requires many queries to be specified. i also like that when i revisit the code, i'll have a more immediate idea of what db-related operations the API class requires

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Looks good. I ended up diving into the code for a couple of hours this morning and dealing with a few different edge-cases and hopefully making it easier to follow. 

@achou11 could you add 3 more tests (see c8 coverage - adding these tests should lead to 100% coverage):

1. addProject throwing should result in accept throwing
2. peer disconnects before reject() is called should not throw
3. peer disconnects before already response is sent should not throw

There is a case that we are not properly handling: when the original invitor disconnects and the invitee tries to accept. I don't think we should automatically accept an invite from the next connected peer, but then not sure what to respond to those peers. I think what should happen is we emit a second invite-received event for the next connected peer, so the user can see that the first invite they got failed, and see a second invite (which may be with a different role). I've written a test for this behaviour, but skipping for now.

I think it's ok to merge without this edge-case being handled. Follow up tasks/issues:

- Handle case (above) if invitor disconnects and other peers have sent invite
- Add role (of invitee) and name (of invitor) to invites

File: src/invite-api.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+    this.#pendingInvites.set(projectId, { fromPeerId: peerId, invite })
+    this.emit('invite-received', {
+      // TODO: Should this be the project public ID since it can be exposed to the client?
+      // Probably would require changing the public methods to accept the public ID
+      // and using the public ID for #invites and #peersToRespondTo keys instead
+      peerId,
+      projectId,
+      projectName: invite.projectInfo?.name,
+    })
```
Comment (achou11) [inline src/invite-api.js:90]: @gmaclennan any thoughts on the question posed here? my understanding is that whatever's emitted here can be surfaced to a client UI. If that's the case, would it be preferable to expose "public" ids instead of or in addition to what's here, or do we let the class consumer (or end-user client) handle that?

**Review by achou11 — COMMENTED** [review]
tests added 👍

File: src/invite-api.js (hunk 3 above)
Comment (achou11) [reply src/invite-api.js:90]: looks like we want this to be the public id so that it can be directly used in other project-related APIs (which accept a public id)

**Review by achou11 — APPROVED** [review]
Thinking this is ready to merge with follow-ups needed.

Created issue for edge case that isn't handled here: https://github.com/digidem/mapeo-core-next/issues/279
